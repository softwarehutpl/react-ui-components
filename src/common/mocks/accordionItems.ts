export default [{
  title: 'Little Prince part 1',
  description: `Once when I was six years old I saw a magnificent picture in a book, called True
    Stories from Nature, about the primeval forest. It was a picture of a boa constrictor
    in the act of swallowing an animal. Here is a copy of the drawing.`,
}, {
  title: 'Little Prince part 2',
  description: `So then I chose another profession, and learned to pilot airplanes. I have flown a
    little over all parts of the world; and it is true that geography has been very useful
    to me. At a glance I can distinguish China from Arizona. If one gets lost in the
    night, such knowledge is valuable.`,
}, {
  title: 'Little Prince part 3',
  description: `"That is a hat." Then I would never talk to that person about boa constrictors, or
    primeval forests, or stars. I would bring myself down to his level. I would talk to
    him about bridge, and golf, and politics, and neckties. And the grownâˆ’up would
    be greatly pleased to have met such a sensible man.`,
}];
