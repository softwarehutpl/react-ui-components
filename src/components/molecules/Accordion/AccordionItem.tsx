import React, { useState } from 'react';
import styled from 'styled-components';

interface AccordionItemProps {
  item: any;
  showIcon?: boolean;
  descriptionBackgroundColor?: string;
  borderWidth?: string;
  noBorder?: boolean;
  bottomBorder?: boolean;
  itemBackgroundColor?: string;
  titleBorderColor?: string;
  titleBackgroundColor?: string;
  titleColor?: string;
  boldTitle?: boolean;
};

interface AccordionItemContainerProps {
  itemBackgroundColor?: string;
  showDescription: boolean;
};

interface ChevronIconProps {
  showDescription: boolean;
};

interface AccordionItemDescriptionProps {
  showDescription: boolean;
  descriptionBackgroundColor?: string;
}

export const AccordionItemContainer = styled.button<AccordionItemContainerProps>`
  width: 100%;
  padding: 10px;
  border: none;
  background-color: ${({ itemBackgroundColor }) => itemBackgroundColor};
  &:focus {
    outline: none;
  };
`;

export const AccordionItemTitle = styled.div<AccordionItemProps>`
  width: 100%;
  padding: 5px;
  background-color: ${({ titleBackgroundColor }) => '#919191' && titleBackgroundColor};
  color: ${({ titleColor }) => titleColor};
  display: flex;
  font-weight: ${({ boldTitle }) => boldTitle ? '700' : '500'};
  justify-content: space-between;
  border: ${({ titleBorderColor, borderWidth, noBorder }) => !noBorder
    && `${borderWidth} solid ${titleBorderColor}`};
  border-bottom: ${({ titleBorderColor, borderWidth, bottomBorder }) => bottomBorder
    && `${borderWidth} solid ${titleBorderColor}`};
`;

export const AccordionItemDescription = styled.div<AccordionItemDescriptionProps>`
  max-height: ${({ showDescription }) => showDescription ? 'auto' : '0' };
  overflow: hidden;
  transition: all 0.2s ease-out;
  background-color: ${({ descriptionBackgroundColor }) => descriptionBackgroundColor};
  padding: ${({ showDescription }) => showDescription ? '10px 0' : '0'};
  display: flex;
  text-align: left;
`;

const ChevronIcon = styled.div<ChevronIconProps>`
  &::before {
    border-style: solid;
    border-width: 1px 1px 0 0;
    content: '';
    display: inline-block;
    height: 0.45em;
    left: 0;
    position: relative;
    top: 0.15em;
    transition: all 0.2s ease-out;
    transform: ${({ showDescription }) => showDescription ? 'rotate(-45deg)' : 'rotate(135deg)'};
    vertical-align: top;
    width: 0.45em;
  }
`;

const AccordionItem = (props: AccordionItemProps) => {
  const [showDesc, handleShowDesc] = useState(false);

  return (
    <AccordionItemContainer {...props} showDescription={showDesc} onClick={() => handleShowDesc(!showDesc)}>
      <AccordionItemTitle {...props} titleColor={props.titleColor}>
        {props.item.title}
        {props.showIcon && <ChevronIcon showDescription={showDesc} />}
      </AccordionItemTitle>
      <AccordionItemDescription showDescription={showDesc}>{props.item.description}</AccordionItemDescription>
    </AccordionItemContainer>
  );
};

const defaultProps = {
  descriptionBackgroundColor: '#fefefe',
  borderWidth: '1px',
  noBorder: true,
  bottomBorder: true,
  showIcon: true,
  itemBackgroundColor: 'transparent',
  titleBackgroundColor: 'transparent',
  titleColor: '#919191',
  titleBorderColor: '#919191',
  boldTitle: false,
};

AccordionItem.defaultProps = defaultProps;

export default AccordionItem;
