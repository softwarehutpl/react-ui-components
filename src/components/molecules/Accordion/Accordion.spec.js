
import React from 'react';
import { create, act } from 'react-test-renderer';
import 'jest-styled-components';
import accordionItems from '../../../common/mocks/accordionItems';
import Accodrion, { AccordionContainer } from './Accordion';
import AccordionItem, { AccordionItemContainer, AccordionItemTitle, AccordionItemDescription } from './AccordionItem';

describe('Accodrion Component', () => {
  it('should render Accodrion correctly', () => {
    const tree = create(<Accodrion items={accordionItems} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render AccodrionContainer correctly', () => {
    const tree = create(<AccordionContainer items={accordionItems} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render AccordionItem correctly', () => {
    const tree = create(<AccordionItem item={accordionItems[0]} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render AccordionItemContainer correctly', () => {
    const tree = create(<AccordionItemContainer item={accordionItems[0]} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render AccordionItemDescription correctly', () => {
    const tree = create(<AccordionItemDescription item={accordionItems[0]} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render AccordionItemTitle correctly', () => {
    const tree = create(<AccordionItemTitle item={accordionItems[0]} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should call button onClick method in AccordionContainer', () => {
    const mockOnChange = jest.fn();
    const component = create(<AccordionItemContainer onClick={mockOnChange} />);
    const instance = component.root;
    const button = instance.findByType('button');
    act(() => {
      button.props.onClick();
    });
    expect(mockOnChange.mock.calls.length).toEqual(1);
  });

  it('should render AccordionContainer with default width', () => {
    const component = create(<AccordionContainer />);
    const accordionContainer = component.toJSON();
    expect(accordionContainer).toHaveStyleRule('width', 'auto');
  });

  it('should render AccordionContainer with introduced width', () => {
    const component = create(<AccordionContainer width={200} />);
    const accordionContainer = component.toJSON();
    expect(accordionContainer).toHaveStyleRule('width', '200px');
  });
  
  it('should render AccordionItemContainer with introduced params', () => {
    const component = create(<AccordionItemContainer itemBackgroundColor='grey' />);
    const accordionItemContainer = component.toJSON();
    expect(accordionItemContainer).toHaveStyleRule('background-color', 'grey');
  });

  it('should render AccordionItemTitle with default params', () => {
    const component = create(<AccordionItemTitle item={accordionItems[0]} />);
    const accordionItemTitle = component.toJSON();
    expect(accordionItemTitle).toHaveStyleRule('padding', '5px');
    expect(accordionItemTitle).toHaveStyleRule('width', '100%');
    expect(accordionItemTitle).toHaveStyleRule('font-weight', '500');
  });
});
