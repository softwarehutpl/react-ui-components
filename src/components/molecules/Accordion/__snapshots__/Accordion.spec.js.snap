// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accodrion Component should render Accodrion correctly 1`] = `
.c1 {
  width: 100%;
  padding: 10px;
  border: none;
  background-color: transparent;
}

.c1:focus {
  outline: none;
}

.c2 {
  width: 100%;
  padding: 5px;
  background-color: transparent;
  color: #919191;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-weight: 500;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  border-bottom: 1px solid #919191;
}

.c4 {
  max-height: 0;
  overflow: hidden;
  -webkit-transition: all 0.2s ease-out;
  transition: all 0.2s ease-out;
  padding: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  text-align: left;
}

.c3::before {
  border-style: solid;
  border-width: 1px 1px 0 0;
  content: '';
  display: inline-block;
  height: 0.45em;
  left: 0;
  position: relative;
  top: 0.15em;
  -webkit-transition: all 0.2s ease-out;
  transition: all 0.2s ease-out;
  -webkit-transform: rotate(135deg);
  -ms-transform: rotate(135deg);
  transform: rotate(135deg);
  vertical-align: top;
  width: 0.45em;
}

.c0 {
  width: auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

<div
  className="c0"
>
  <button
    className="c1"
    onClick={[Function]}
  >
    <div
      className="c2"
    >
      Little Prince part 1
      <div
        className="c3"
      />
    </div>
    <div
      className="c4"
    >
      Once when I was six years old I saw a magnificent picture in a book, called True
    Stories from Nature, about the primeval forest. It was a picture of a boa constrictor
    in the act of swallowing an animal. Here is a copy of the drawing.
    </div>
  </button>
  <button
    className="c1"
    onClick={[Function]}
  >
    <div
      className="c2"
    >
      Little Prince part 2
      <div
        className="c3"
      />
    </div>
    <div
      className="c4"
    >
      So then I chose another profession, and learned to pilot airplanes. I have flown a
    little over all parts of the world; and it is true that geography has been very useful
    to me. At a glance I can distinguish China from Arizona. If one gets lost in the
    night, such knowledge is valuable.
    </div>
  </button>
  <button
    className="c1"
    onClick={[Function]}
  >
    <div
      className="c2"
    >
      Little Prince part 3
      <div
        className="c3"
      />
    </div>
    <div
      className="c4"
    >
      "That is a hat." Then I would never talk to that person about boa constrictors, or
    primeval forests, or stars. I would bring myself down to his level. I would talk to
    him about bridge, and golf, and politics, and neckties. And the grownâˆ’up would
    be greatly pleased to have met such a sensible man.
    </div>
  </button>
</div>
`;

exports[`Accodrion Component should render AccodrionContainer correctly 1`] = `
.c0 {
  width: auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

<div
  className="c0"
/>
`;

exports[`Accodrion Component should render AccordionItem correctly 1`] = `
.c0 {
  width: 100%;
  padding: 10px;
  border: none;
  background-color: transparent;
}

.c0:focus {
  outline: none;
}

.c1 {
  width: 100%;
  padding: 5px;
  background-color: transparent;
  color: #919191;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-weight: 500;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  border-bottom: 1px solid #919191;
}

.c3 {
  max-height: 0;
  overflow: hidden;
  -webkit-transition: all 0.2s ease-out;
  transition: all 0.2s ease-out;
  padding: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  text-align: left;
}

.c2::before {
  border-style: solid;
  border-width: 1px 1px 0 0;
  content: '';
  display: inline-block;
  height: 0.45em;
  left: 0;
  position: relative;
  top: 0.15em;
  -webkit-transition: all 0.2s ease-out;
  transition: all 0.2s ease-out;
  -webkit-transform: rotate(135deg);
  -ms-transform: rotate(135deg);
  transform: rotate(135deg);
  vertical-align: top;
  width: 0.45em;
}

<button
  className="c0"
  onClick={[Function]}
>
  <div
    className="c1"
  >
    Little Prince part 1
    <div
      className="c2"
    />
  </div>
  <div
    className="c3"
  >
    Once when I was six years old I saw a magnificent picture in a book, called True
    Stories from Nature, about the primeval forest. It was a picture of a boa constrictor
    in the act of swallowing an animal. Here is a copy of the drawing.
  </div>
</button>
`;

exports[`Accodrion Component should render AccordionItemContainer correctly 1`] = `
.c0 {
  width: 100%;
  padding: 10px;
  border: none;
}

.c0:focus {
  outline: none;
}

<button
  className="c0"
/>
`;

exports[`Accodrion Component should render AccordionItemDescription correctly 1`] = `
.c0 {
  max-height: 0;
  overflow: hidden;
  -webkit-transition: all 0.2s ease-out;
  transition: all 0.2s ease-out;
  padding: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  text-align: left;
}

<div
  className="c0"
/>
`;

exports[`Accodrion Component should render AccordionItemTitle correctly 1`] = `
.c0 {
  width: 100%;
  padding: 5px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-weight: 500;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  border: undefined solid undefined;
}

<div
  className="c0"
/>
`;
